src/plugins/intel_npu/src/plugin/npuw/util.cpp

```diff
@@ -1052,6 +1048,159 @@ void unpack_u4f16_asymm_zp(const ov::SoPtr<ov::ITensor>& from,
+        uint8_t const* pZerLocal = pZer + zerop_elem_type.size() * sindex * stride / 2;
+        int16_t* pDstLocal = const_cast<int16_t*>(pDst) + 64 * elementsPerScaleVectorized * sindex * stride;
+
+        // if it is last iteration current stride can be smaller - lets check that
+        sindex *= stride;
+        const auto jobFinish = std::min(sindex + stride, stotal);
+
+        for (; sindex < jobFinish; sindex++) {
+            __m256 svalVec = avx2_load_scale(pSclLocal, scale_elem_type);
+            __m256 zvalVec = _mm256_set1_ps(static_cast<float>(*pZerLocal));
+
+            for (std::size_t index = 0; index < elementsPerScale; index += 64) {
+                __m128i* outv[] = {
+                    reinterpret_cast<__m128i*>(pDstLocal),
+                    reinterpret_cast<__m128i*>(pDstLocal + 8),
+                    reinterpret_cast<__m128i*>(pDstLocal + 16),
+                    reinterpret_cast<__m128i*>(pDstLocal + 24),
+                    reinterpret_cast<__m128i*>(pDstLocal + 32),
+                    reinterpret_cast<__m128i*>(pDstLocal + 40),
+                    reinterpret_cast<__m128i*>(pDstLocal + 48),
+                    reinterpret_cast<__m128i*>(pDstLocal + 56),
+                };
+
+                __m128i xmmUnpacked1 = _mm_loadu_si128(reinterpret_cast<__m128i const*>(pSrcLocal));
+                __m128i xmmUnpacked2 = _mm_loadu_si128(reinterpret_cast<__m128i const*>(pSrcLocal + 16));
+                __m128i xmmUnpacked3 = _mm_loadu_si128(reinterpret_cast<__m128i const*>(pSrcLocal + 32));
+                __m128i xmmUnpacked4 = _mm_loadu_si128(reinterpret_cast<__m128i const*>(pSrcLocal + 48));
+
+                // converting to 32 x f16
+                __m128i f16LoLo[] = {avx2_u8tof16_hi(xmmUnpacked1, zvalVec, svalVec),
+                                     avx2_u8tof16_lo(xmmUnpacked1, zvalVec, svalVec)};
+
+                __m128i f16LoHi[] = {
+                    avx2_u8tof16_hi(xmmUnpacked2, zvalVec, svalVec),
+                    avx2_u8tof16_lo(xmmUnpacked2, zvalVec, svalVec),
+                };
+
+                __m128i f16HiLo[] = {avx2_u8tof16_hi(xmmUnpacked3, zvalVec, svalVec),
+                                     avx2_u8tof16_lo(xmmUnpacked3, zvalVec, svalVec)};
+                __m128i f16HiHi[] = {avx2_u8tof16_hi(xmmUnpacked4, zvalVec, svalVec),
+                                     avx2_u8tof16_lo(xmmUnpacked4, zvalVec, svalVec)};
+
+                // store the results
+                _mm_storeu_si128(outv[0], f16LoLo[0]);
+                _mm_storeu_si128(outv[1], f16LoLo[1]);
+                _mm_storeu_si128(outv[2], f16LoHi[0]);
+                _mm_storeu_si128(outv[3], f16LoHi[1]);
+                _mm_storeu_si128(outv[4], f16HiLo[0]);
+                _mm_storeu_si128(outv[5], f16HiLo[1]);
+                _mm_storeu_si128(outv[6], f16HiHi[0]);
+                _mm_storeu_si128(outv[7], f16HiHi[1]);
```

Comment:
Accuracy is bad. Probably something is wrong with intrinsics here

Reply:
Start with a scalar code then

Reply:
Already wip, just a note for self :)

Reply:
Should be working now. However there are other accuracy issues related to the model